galsim/chromatic.py

```diff
@@ -1223,6 +1259,50 @@ def evaluateAtWavelength(self, wave):
         im, stepk, maxk = self._imageAtWavelength(wave)
         return InterpolatedImage(im, _force_stepk=stepk, _force_maxk=maxk)
 
+    def _shoot(self, photons, rng):
+        from .photon_array import PhotonArray
+        from .random import UniformDeviate
+
+        w = photons.wavelength
+        if np.any((w < self.waves[0]) | (w > self.waves[-1])):
+            raise GalSimRangeError("Shooting photons outside the interpolated wave_list",
+                                   w, self.waves[0], self.waves[-1])
+
+        k = np.searchsorted(self.waves, w)
+        k[k==0] = 1  # if k == 0, then w == min(waves). Using k=1 instead is fine for this.
+        #assert np.all(k > 0)
+        #assert np.all(k < len(self.waves))
+
+        # For each w, these are the wavelengthat that bracket w:
+        w0 = self.waves[k-1]
+        w1 = self.waves[k]
+        #assert np.all(w0 <= w)
+        #assert np.all(w <= w1)
+
+        # If we could get away with averaging photons shot at each wavelength,
+        # these would be relative fractions.  So e.g. x = x0 f0 + x1 f1 would be the
+        # right weighted average to use.
+        f0 = (w1-w) / (w1-w0)
+        #f1 = (w-w0) / (w1-w0)  (We don't need this quantity below.)
+
+        # Instead of averaging these, we can do the averaging probabilistically by selecting
+        # each photon with a probability equal to the relative weight we would have used
+        # in the average.
+        u = np.empty(len(photons))
+        UniformDeviate(rng).generate(u)
+        use_k = k - (u<f0).astype(int)  # The second term is either 0 or 1.
+
+        # Draw photons from the saved profiles according to when we have selected to use each one.
+        for kk, ww in enumerate(self.waves):
+            use = use_k == kk
```

Comment:
I think `use = (use_k == kk)` is clearer.  I missed the double equals the first couple times I read this line.